### PR TITLE
add testcase for validating rss hash-key change

### PR DESCRIPTION
**Pull request by sharsonia**

### CODE REVIEW

File: lisa/tools/ethtool.py [hunk 1]
```diff
@@ -499,6 +544,42 @@ def change_device_ring_buffer_settings(
...
+    def change_device_rss_hash_key(
+        self, interface: str, hash_key: str
+    ) -> DeviceRssHashKey:
+        result = self.run(f"-X {interface} hkey {hash_key}", sudo=True, force_run=True)
+        if (result.exit_code != 0) and ("Operation not supported" in result.stdout):
+            raise UnsupportedOperationException(
+                f"Changing RSS hash key with 'ethtool -X {interface}' not supported."
+            )
+        result.assert_exit_code(
+            message=f" Couldn't change device {interface} hash key."
```
Comment (squirrelsc) [inline lisa/tools/ethtool.py:617]: remove the extra whitespace at the beginning.

File: microsoft/testsuites/network/networksettings.py [hunk 2]
```diff
@@ -264,3 +264,57 @@ def validate_device_gro_lro_settings_change(self, node: Node, log: Logger) -> No
...
+
+            Steps:
+            1. Get all the device's RSS hash key values.
+            2. Try flipping the GRO and LRO settings and validate it takes affect.
+            3. Revert back the settings to original values.
+        """,
+        priority=2,
+    )
+    def validate_device_rss_hash_key_change(self, node: Node, log: Logger) -> None:
+        if node.os.information.version.compare("5.0.0") < 0:
```
Comment (squirrelsc) [inline microsoft/testsuites/network/networksettings.py]: Try
`if node.os.information.version < "5.0.0":`

File: microsoft/testsuites/network/networksettings.py [hunk 3]
```diff
@@ -264,3 +264,59 @@ def validate_device_gro_lro_settings_change(self, node: Node, log: Logger) -> No
...
+            raise SkippedException(
+                f"The kernel version {node.os.information.version} does not support"
+                " changing RSS hash key."
+            )
+
+        ethtool = node.tools[Ethtool]
+        try:
+            devices_rss_hkey_info = ethtool.get_all_device_rss_hash_key()
+        except UnsupportedOperationException as identifier:
+            raise SkippedException(identifier)
```
Comment (squirrelsc) [inline microsoft/testsuites/network/networksettings.py:310]: Make sure the error message is human readable.

File: microsoft/testsuites/network/networksettings.py [hunk 4]
```diff
@@ -264,3 +264,59 @@ def validate_device_gro_lro_settings_change(self, node: Node, log: Logger) -> No
...
+            # Swap the last 2 characters of the original hash key to make new hash key.
+            split_hkey = original_hkey.rsplit(":", 1)
+            swapped_part = "".join(
+                [
+                    split_hkey[1][x : x + 2][::-1]
+                    for x in range(0, len(split_hkey[1]), 2)
+                ]
+            )
+
+            new_hkey = split_hkey[0] + ":" + swapped_part
```
Comment (squirrelsc) [inline microsoft/testsuites/network/networksettings.py]: Use f string to format it. `f"{split_hkey[0]}:{swapped_part}"`

Comment (squirrelsc) [reply microsoft/testsuites/network/networksettings.py]: call the variable `expected_key`. It's clearer in test case, expected v.s. actual.

File: lisa/tools/ethtool.py [hunk 5]
```diff
@@ -259,6 +259,47 @@ def _parse_gro_lro_settings_info(self, interface: str, raw_str: str) -> None:
...
+
+    _rss_hash_key_pattern = re.compile(
+        r"^RSS hash key:[\s+](?P<value>.*?)?$", re.MULTILINE
+    )
+
+    def __init__(self, interface: str, device_rss_hash_info_raw: str) -> None:
+        self._parse_rss_hash_key(interface, device_rss_hash_info_raw)
+
+    def _parse_rss_hash_key(self, interface: str, raw_str: str) -> None:
+        hash_key_pattern = self._rss_hash_key_pattern.search(raw_str)
```
Comment (LiliDeng) [inline lisa/tools/ethtool.py:293]: just hash_key is enough?

File: microsoft/testsuites/network/networksettings.py [hunk 6]
```diff
@@ -264,3 +264,59 @@ def validate_device_gro_lro_settings_change(self, node: Node, log: Logger) -> No
...
+            1. Skip the test if the kernel version is any less than LTS 5.
+            2. Get all the device's RSS hash key values.
+            3. Swap the last 2 characters of original hash key to make a new hash key.
+            4. Validate changing the hash key setting using the new hash key.
+            5. Revert back the settings to original values.
+        """,
+        priority=2,
+    )
+    def validate_device_rss_hash_key_change(self, node: Node, log: Logger) -> None:
+        if node.os.information.version < "5.0.0":
```
Comment (LiliDeng) [inline microsoft/testsuites/network/networksettings.py]: Here, I have some concern, this condition may good for debian/ubuntu distro.
But the highest kernel version for rhel/centos is 4.X, I think it supports change hkey.

File: microsoft/testsuites/network/networksettings.py [hunk 7]
```diff
@@ -264,3 +266,66 @@ def validate_device_gro_lro_settings_change(self, node: Node, log: Logger) -> No
...
+            4. Validate changing the hash key setting using the new hash key.
+            5. Revert back the settings to original values.
+        """,
+        priority=2,
+    )
+    def validate_device_rss_hash_key_change(self, node: Node, log: Logger) -> None:
+        if isinstance(node.os, Debian):
+            min_supported_kernel = "5.0.0"
+        elif isinstance(node.os, Redhat):
+            min_supported_kernel = "4.0.0"
```
Comment (squirrelsc) [inline microsoft/testsuites/network/networksettings.py:291]: add `else` to handle other distros.

File: microsoft/testsuites/network/networksettings.py [hunk 8]
```diff
@@ -264,3 +266,74 @@ def validate_device_gro_lro_settings_change(self, node: Node, log: Logger) -> No
...
+    def validate_device_rss_hash_key_change(self, node: Node, log: Logger) -> None:
+        uname = node.tools[Uname]
+        linux_info = uname.get_linux_information()
+
+        if isinstance(node.os, Debian):
+            min_supported_kernel = "5.0.0"
+        elif isinstance(node.os, Redhat):
+            min_supported_kernel = "4.0.0"
+        elif isinstance(node.os, Suse):
+            min_supported_kernel = "4.12.14-120.1"
```
Comment (squirrelsc) [inline microsoft/testsuites/network/networksettings.py]: The `-120.1` doesn't support comparison by the VersionInfo. Is it needed?

File: lisa/tools/ethtool.py [hunk 9]
```diff
@@ -259,6 +259,57 @@ def _parse_gro_lro_settings_info(self, interface: str, raw_str: str) -> None:
...
+        self.rss_hash_key = hash_key_pattern.group("value")
+
+
+class DeviceRxHashLevel:
+    def __init__(
+        self, interface: str, protocol: str, device_rx_hash_level_raw: str
+    ) -> None:
+        self._parse_rx_hash_level(interface, protocol, device_rx_hash_level_raw)
+
+    def _parse_rx_hash_level(self, interface: str, protocol: str, raw_str: str) -> None:
```
Comment (squirrelsc) [inline lisa/tools/ethtool.py:324]: There is no value saved. Is it useful?

**Comment by sharsonia** [comment]
Results for validate_device_rx_hash_level_change test case

[lisa-20210831-220017-247.log](https://github.com/microsoft/lisa/files/7087018/lisa-20210831-220017-247.log)

**Review by squirrelsc — APPROVED** [review]

**Pull request merged**